File: 016__3Sum_Closest.py
class Solution(object):
    def threeSumClosest(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        nums.sort()
        min_val = float("inf")
        for i in range(len(nums) - 2):
            if i > 0 and nums[i] == nums[i - 1]:
                continue;
            l, r = i + 1, len(nums) - 1
            while l < r:
                val = nums[i] + nums[l] + nums[r]
                if val == target:
                    return target
                elif val > target:
                    if abs(val - target) < abs(min_val - target):
                        min_val = val
                    r -= 1
                else:
                    if abs(val - target) < abs(min_val - target):
                        min_val = val
                    l += 1
        return min_val

File: Java/016__3Sum_Closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {

        int min_val = nums[0] + nums[1] + nums[2];
        Arrays.sort(nums);
        for (int i = 0; i < nums.length - 2; i++) {
            int l = i + 1, r = nums.length - 1;
            while (l < r) {
                int sum = nums[i] + nums[l] + nums[r];
                if (Math.abs(sum - target) < Math.abs(min_val - target)) {
                    min_val = sum;
                }
                if (sum - target > 0) {
                    r--;
                } else {
                    l++;
                }
            }
        }
        return min_val;
    }
}
